### ISSUE TITLE
Opening a (Possibly Malicious) Link Crashes Sway

### ISSUE THREAD

**Issue by ghost** [swaywm/sway#6696]
- **Sway Version:**
  - `sway version 1.6.1`

- **Description:**
  - (CAUTION: THIS MAY CRASH SWAY.) When I open https://github.com/j34nd/Y4J4N on FIrefox, Sway immediately crashes. There's some random hex data in that repo. I suspect it might be malicious. (I am not a security expert.) Please look into it.

**Comment by vyivel** [comment]
The crash is presumably caused by a very big xdg_popup which is created when you hover over the tab. Mutter and KWin seem to handle this case just fine, without any crashes. On my machine, only Firefox crashed, without Sway.

**Comment by aragon999** [comment]
Using Firefox 94 I also noticed a crash of Firefox (and sway) when opening a symfony debug error page, with a lot of content.

As @vyivel suggested I think that it is indeed related to the amount of data which is on the page, since the crash only happens if there is a large stack trace, and works if it is smaller (the debug page always includes the full source code of each step, so it can get quite large).
So I think this is a Firefox bug (introduced somewhere at Firefox version 92/93), but I had no time to investigate this issue further.

For me sway crashes (as well as it does on your link) with the error:
```
sway: render/wlr_texture.c:29: wlr_texture_from_pixels: Assertion `data' failed.
```

But I think that is related to the nouveau driver, the stack trace reads:
<details>
    <summary>stack-trace-without-debug.log</summary>

```
systemd-coredump[9221]: [🡕] Process 798 (sway) of user 1000 dumped core.
    Found module linux-vdso.so.1 with build-id: 21f54a7973875d5754f43db6afff2d13e5604665
    Found module libxshmfence.so.1 with build-id: 8876d9ccf620858795724ca24b9e567585a77cec
    Found module libxcb-sync.so.1 with build-id: dda14591103b01b1311906053bf1ca9e82ade35c
    Found module libxcb-dri2.so.0 with build-id: 2dd6e65129a809dab828a1d26215a3f7a363fcc8
    Found module libX11-xcb.so.1 with build-id: 0db4f94d8ae31b8dc9a83f825a9171656f1e532c
    Found module libEGL_mesa.so.0 with build-id: fb943970628f1b739095eb6e3a22e90b402ecd0a
    Found module libnvidia-egl-gbm.so.1 with build-id: f5eda6dba29b7a866131b070d4953ffbb41fbc4c
    Found module libnvidia-eglcore.so.495.44 with build-id: 8bdda01a4252b83581c7f6563a0ae5bcf0ac1d8d
    Found module libnvidia-egl-wayland.so.1 with build-id: 7026a2a5082aff391d23ed5af4678c543383f2d9
    Found module libnvidia-glsi.so.495.44 with build-id: 224ed2e0e9fdbb44ecbe3884cdedde88fcb509ba
    Found module libEGL_nvidia.so.0 with build-id: 7df525a19bd457c97173afa74db3c1d8a69f78a6
    Found module libnss_files.so.2 with build-id: 1a36dfc01d3a1010b2ee79766a24a8090a3266d5
    Found module libicudata.so.70 with build-id: e1dcc2a88cfaafed882d09c90c668af0eed4efed
    Found module libicuuc.so.70 with build-id: 2e245c2bf12f95fd8ab79b3a4be99524677cbd70
    Found module libxml2.so.2 with build-id: 6e67981c9b7528a3502824bf183423a0a02e1075
    Found module libncursesw.so.6 with build-id: 859adb8b03a5655151b4da104bc4b326a0d4bc8c
    Found module libedit.so.0 with build-id: 4b0babfcad161c2ad0af6e59e2493258db23a331
    Found module libgcc_s.so.1 with build-id: 7f8508bb914546ada778809b64b99d234337d835
    Found module libstdc++.so.6 with build-id: 8ab0e57054dd1dcba681f217016afc6a4e639783
    Found module libvulkan.so.1 with build-id: 3d98807d94dc94c0f97c89da76c24bba96a6c46c
    Found module libdrm_nouveau.so.2 with build-id: ee16fa6a4d0bd08f5448cfc046ccb76b49aaa8a3
    Found module libdrm_amdgpu.so.1 with build-id: 30d019987d514b23115b47d6b47c1e45c18fdbaf
    Found module libelf.so.1 with build-id: 4300a0a24ac790c27ac0fe1fd5253729905ce708
    Found module libdrm_radeon.so.1 with build-id: 5d1615e0467d0a34a5a82cbd4ebc84851e116307
    Found module libsensors.so.5 with build-id: dc8b2c1c0d8525411aca188ea3cc3fb86d381d30
    Found module libLLVM-13.so with build-id: 82982c299d46219d0eab8ee9cb8bb7365483453b
    Found module nouveau_dri.so with build-id: 485271131aae5a573f803418b6bbb1f58d034acb
    Found module libglapi.so.0 with build-id: 57828fb70af7c1f2aa70b096d748851a1fd39a9c
    Found module libgpg-error.so.0 with build-id: 3801187d42c4955bd93c605451430cdf1b164e63
    Found module libgcrypt.so.20 with build-id: db45f5d5e0f7af1e77324fea1885f974619ad268
    Found module libcap.so.2 with build-id: 7ab63cdf1c7435ae6910ccd518a92dc4f6a9b5b6
    Found module liblz4.so.1 with build-id: e63600ab23b2f6997f42fac2fa56e1f02ce159a1
    Found module libzstd.so.1 with build-id: 4b10444c1560ebc574af4d5f488b7408b22d450e
    Found module liblzma.so.5 with build-id: 8b615460aa230708c5183f16bede67aa0437d95e
    Found module libblkid.so.1 with build-id: f03a24e92c9852272bd0049b5bcb7ed6cf0e664f
    Found module libbrotlicommon.so.1 with build-id: f68934f94312f770550ebc6c5acfd6359b098b07
    Found module libsystemd.so.0 with build-id: efcf1b39ffd5f65334b94df717072c7e7a1b5fa0
    Found module libgraphite2.so.3 with build-id: 47761dc11e553f519cde97ed9ee985be12ccdae2
    Found module libdatrie.so.1 with build-id: 6fe3b6ece2c8e7d11869fa051375128d8f808f58
    Found module libresolv.so.2 with build-id: c915c72668282861a813f7ea3c0780f37b681dc0
    Found module libmount.so.1 with build-id: ff9a1d99d35cf640d0bfdaba294854372672c29b
    Found module libgmodule-2.0.so.0 with build-id: 34cabbd12d9baf397a0255f174e4c96165bc329b
    Found module libgudev-1.0.so.0 with build-id: a9f734ea9206d637d5aacfbd86298c91cdc1a33a
    Found module libdl.so.2 with build-id: 5abc547e7b0949f89f3c0e21ab0c8331a7440a8a
    Found module libbrotlidec.so.1 with build-id: f871e6f204ab8d48099915126ba01be989a8000f
    Found module libbz2.so.1.0 with build-id: 919597c477c9b2cb9cdbb7745ed6494ac0e6da60
    Found module libexpat.so.1 with build-id: 8850138eae6d9d4d43c5c4b2ac48393bc4279037
    Found module libpangoft2-1.0.so.0 with build-id: 0a64cffe87210d8b7864a3a4fdaa92a120d20bee
    Found module libXdmcp.so.6 with build-id: 8ca0792d23c8b8b4c0864297512349292bea5955
    Found module libXau.so.6 with build-id: 1c67764663e07bec24d8951e5fd93f4d165979ff
    Found module libxcb-errors.so.0 with build-id: 0d6ee255fc30b6a86856b9f1a96491f12a1cacaf
    Found module libxcb-res.so.0 with build-id: b825f66a75478fd00eeb4c57ecb0ac7792e00f5e
    Found module libxcb-icccm.so.4 with build-id: 73f1e12d76af383aa969a60279ed5ed922045f47
    Found module libxcb-composite.so.0 with build-id: 94fff40466ef37eb5973997211aa68f9beb81d15
    Found module libseat.so.1 with build-id: 08086618ff09f1adf85ac58eab286c3435c952db
    Found module libxcb-xinput.so.0 with build-id: 6403174c48f05f40cf7c168beb86f2886b9f09c1
    Found module libxcb-xfixes.so.0 with build-id: a6b197ace5b9b59f913f5969eb419a88d1194f47
    Found module libxcb-render-util.so.0 with build-id: c64fb28ea20a320a9754273d9ea3af85f6f74b17
    Found module libxcb-present.so.0 with build-id: 68f5465258750e2397b1333b3ffc01ee33caa4e1
    Found module libxcb-dri3.so.0 with build-id: 9407a2480e09dc5a1dd9d9a0652fa8d32b328c91
    Found module libEGL.so.1 with build-id: 473944aea0e3dbc5d595dee647702a3cdc0aed78
    Found module libgbm.so.1 with build-id: 42e129ba16de8ff426f740e7686a1f02a738abf9
    Found module libdrm.so.2 with build-id: 457a1a1bd25f1d5541a10d3f16c5dc1440d77c7d
    Found module libwayland-client.so.0 with build-id: d274361e7aa57c9b63504a039e063322450cbb9c
    Found module libGLdispatch.so.0 with build-id: 195b34c1bb271f3b1162c897cfbcfb859e656b93
    Found module libffi.so.8 with build-id: f90d8b734f6de9b25faedb8cbfab7054dafc0a42
    Found module libharfbuzz.so.0 with build-id: 5d9b408e95d3c98a95333fdd413b8b64a7757631
    Found module libthai.so.0 with build-id: 3fdce1a7db2b94d9168eea540dbf5ff5caa3b8ad
    Found module libfribidi.so.0 with build-id: 79124bad061c2aab4ec89f25d067363d5b781114
    Found module libgio-2.0.so.0 with build-id: b3348994e4d526d182becc32821aa3dbe9797149
    Found module librt.so.1 with build-id: 75484da2d6f1515189eefa076e0a40328834cd16
    Found module libwacom.so.2 with build-id: 1d9b89a6d39be5568d5f44888b8a74f46a30c152
    Found module libmtdev.so.1 with build-id: 0cea2a842ae9a0ef02a08477076caab0e5ac5e42
    Found module ld-linux-x86-64.so.2 with build-id: 040cc3dd10461562f177df39e3be2f3704258c3c
    Found module libxcb-shm.so.0 with build-id: fb797f299a446f559a95afcc168227482cc800d1
    Found module libxcb-render.so.0 with build-id: a37bdb37744b508be9dd29fd9ccb9bb0170d43e5
    Found module libXrender.so.1 with build-id: 97e0b9ab6ba96ebc86527cc2b3c3078aad8616b3
    Found module libXext.so.6 with build-id: d70f24beb4fad748d6becffdcc13e51be0a2ebfa
    Found module libX11.so.6 with build-id: e9192497aa39fab51358826d1a1b5fb77edc5081
    Found module libfreetype.so.6 with build-id: 26c5f833068ff72660d1975cbc2074c3eb47fad8
    Found module libfontconfig.so.1 with build-id: 1103a641395c7d3b42e49b793d3a9ea927c77bf6
    Found module libpng16.so.16 with build-id: 2dc0bce07f199bf983c07a05fb95a6f4af83a9b3
    Found module libz.so.1 with build-id: 81bf6e728a6d6f5b105b0f8b25f6c614ce10452a
    Found module libpthread.so.0 with build-id: 07c8f95b4f3251d08550217ad8a1f31066229996
    Found module libc.so.6 with build-id: 4b406737057708c0e4c642345a703c47a61c73dc
    Found module libpangocairo-1.0.so.0 with build-id: 820de243a3ad1d1d9e12ebc2b6073b19d620d744
    Found module libxcb.so.1 with build-id: 0d1ef11740a5daad2ee331e812a51aa6574af222
    Found module libxkbcommon.so.0 with build-id: 1f1bc7527e57c886c3df5803068122e1971d4724
    Found module libwlroots.so.9 with build-id: b0a3eaf1e7b40cf9b49677901af2dc7f22b718b2
    Found module libwayland-server.so.0 with build-id: 7156897eafc4b7562f2b214e6ad8ca6c85c57e47
    Found module libpixman-1.so.0 with build-id: 341f793dcada3a48a306a793d265a517e3f2e7d6
    Found module libGLESv2.so.2 with build-id: a56281a2afd5eea8cd258ea2487094cd322e8c7e
    Found module libpcre.so.1 with build-id: 845483dd0acba86de9f0313102bebbaf3ce52767
    Found module libglib-2.0.so.0 with build-id: d4aa102e545015a506c3ee06398676b70f65bcbc
    Found module libgobject-2.0.so.0 with build-id: 3e451709a0a5772b9550d3a4c28dcb37a87028fa
    Found module libpango-1.0.so.0 with build-id: 0036caf67265c11b1de7a50321d9bee619248118
    Found module libm.so.6 with build-id: 2b8fd1f869ecab4e0b55e92f2f151897f6818acf
    Found module libudev.so.1 with build-id: 996dbb21212808b5930dd015512fcdd760d4cf1a
    Found module libinput.so.10 with build-id: 77f3eca5dddc8cac29b71bc5041a79ce57fceb56
    Found module libevdev.so.2 with build-id: b62ae69f839f1b6b06ef2c1df19f25ff09b0d824
    Found module libjson-c.so.5 with build-id: 58a98db7dcfc01670a5565a0cc89c405d82b4a09
    Found module libcairo.so.2 with build-id: 06820af9cf79c2deb207d9533a61caf04e67a8ad
    Found module sway with build-id: 827352ff3c70254c772ba986f7021f1bf425c31e
    Stack trace of thread 798:
    #0  0x00007f31cebccd22 raise (libc.so.6 + 0x3cd22)
    #1  0x00007f31cebb6862 abort (libc.so.6 + 0x26862)
    #2  0x00007f31cebb6747 __assert_fail_base.cold (libc.so.6 + 0x26747)
    #3  0x00007f31cebc5616 __assert_fail (libc.so.6 + 0x35616)
    #4  0x00007f31cee07443 wlr_texture_from_pixels (libwlroots.so.9 + 0x27443)
    #5  0x000055c3ab02f885 n/a (sway + 0x52885)
    #6  0x000055c3ab032d0e n/a (sway + 0x55d0e)
    #7  0x000055c3ab03ec8f n/a (sway + 0x61c8f)
    #8  0x000055c3aaffba62 n/a (sway + 0x1ea62)
    #9  0x00007f31cee531ee wlr_signal_emit_safe (libwlroots.so.9 + 0x731ee)
    #10 0x00007f31ce59bd4a n/a (libffi.so.8 + 0x6d4a)
    #11 0x00007f31ce59b267 n/a (libffi.so.8 + 0x6267)
    #12 0x00007f31ceea6173 n/a (libwayland-server.so.0 + 0xd173)
    #13 0x00007f31ceea155c n/a (libwayland-server.so.0 + 0x855c)
    #14 0x00007f31ceea404a wl_event_loop_dispatch (libwayland-server.so.0 + 0xb04a)
    #15 0x00007f31ceea1bc7 wl_display_run (libwayland-server.so.0 + 0x8bc7)
    #16 0x000055c3aafed97b n/a (sway + 0x1097b)
    #17 0x00007f31cebb7b25 __libc_start_main (libc.so.6 + 0x27b25)
    #18 0x000055c3aafedcee n/a (sway + 0x10cee)
    
    Stack trace of thread 9219:
    #0  0x00007f31cead7abc n/a (libfontconfig.so.1 + 0xaabc)
    #1  0x00007f31cead8311 FcCharSetIsSubset (libfontconfig.so.1 + 0xb311)
    #2  0x00007f31cead970c FcCharSetMerge (libfontconfig.so.1 + 0xc70c)
    #3  0x00007f31ceaf3f3b FcFontSetSort (libfontconfig.so.1 + 0x26f3b)
    #4  0x00007f31ce40331e n/a (libpangoft2-1.0.so.0 + 0x831e)
    #5  0x00007f31cf066ce1 n/a (libglib-2.0.so.0 + 0x82ce1)
    #6  0x00007f31ceb76259 start_thread (libpthread.so.0 + 0x9259)
    #7  0x00007f31cec8e5e3 __clone (libc.so.6 + 0xfe5e3)
    
    Stack trace of thread 865:
    #0  0x00007f31ceb828ca __futex_abstimed_wait_common64 (libpthread.so.0 + 0x158ca)
    #1  0x00007f31ceb7c270 pthread_cond_wait@@GLIBC_2.3.2 (libpthread.so.0 + 0xf270)
    #2  0x00007f31cc55388c n/a (nouveau_dri.so + 0x1c888c)
    #3  0x00007f31cc54cf68 n/a (nouveau_dri.so + 0x1c1f68)
    #4  0x00007f31ceb76259 start_thread (libpthread.so.0 + 0x9259)
    #5  0x00007f31cec8e5e3 __clone (libc.so.6 + 0xfe5e3)
    
    Stack trace of thread 888:
    #0  0x00007f31ceb828ca __futex_abstimed_wait_common64 (libpthread.so.0 + 0x158ca)
    #1  0x00007f31ceb7c270 pthread_cond_wait@@GLIBC_2.3.2 (libpthread.so.0 + 0xf270)
    #2  0x00007f31cc55388c n/a (nouveau_dri.so + 0x1c888c)
    #3  0x00007f31cc54cf68 n/a (nouveau_dri.so + 0x1c1f68)
    #4  0x00007f31ceb76259 start_thread (libpthread.so.0 + 0x9259)
    #5  0x00007f31cec8e5e3 __clone (libc.so.6 + 0xfe5e3)
```
</details>

**tl/dr:** I think it is not a `sway` issue, but rather two issues, one in the nouveau driver and one in Firefox.

**Comment by FreeFull** [comment]
I suspect this might be the same bug as #6531 ? I don't get a crash on my end, but the firefox title renders as a black rectangle for me when on that page. Also, the page's <title> tag seems to contain 29864 characters, which definitely seems more than long enough to trigger that bug.

**Comment by aragon999** [comment]
It very well could be that for me this is the same issue. I will check that, maybe only coincidentally the title was longer for the errors with the large stack trace :-)

I will also try if this commit fixes the issue: 
https://github.com/FlexW/sway/commit/a2763262d41f32e1aa2131c67fbe80a6a43c40df

**Comment by aragon999** [comment]
So I can confirm that https://github.com/swaywm/sway/pull/6556 fixes the problem for the link from this issue. So I think this can be closed as a duplicate.

Thank you :-)
